### PR TITLE
add HTTP error helper

### LINKED ISSUE #331: Provide function to write RPC errors from HTTP middleware
**Is your feature request related to a problem? Please describe.**
Even though Connect handlers work with `net/http` middleware, it's often necessary to re-implement existing HTTP middleware as Connect interceptors so that clients get nicely-annotated errors. This is a little silly.

**Describe the solution you'd like**
It'd be nice for Connect to include a function that HTTP middleware can use to encode Connect errors. `func WriteError(http.ResponseWriter, *http.Request, error)` might work.

**Describe alternatives you've considered**
The status quo is okay, but forces users to jump through more hoops that we'd like.

### PR DESCRIPTION
**Pull request by rhbuf**
The mapping from connect error codes to HTTP status codes is currently private and not accessible to users who are implementing connect in their existing HTTP servers. This helper has been added to allow those users to easily update existing HTTP middlewares to handle connect errors.

fixes #331

### CODE REVIEW

File: http_util.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+package connect
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+// WriteError should be called from HTTP middlewares to handle connect errors. It
+// writes a status code encodes a connect error as JSON,  and writes to the response.
+func WriteError(responseWriter http.ResponseWriter, err error) error {
```
Comment (rhbuf) [inline http_util.go:24]: @akshayjshah originally had a request in here, but I didn't see any reason to include it. Potentially it was there to close the request body, and when I lifted this code from https://github.com/bufbuild/connect-go/blob/main/protocol_connect.go#L553 I did see that function closing the request. But when looking into https://pkg.go.dev/net/http#Request.Close I believe that `WriteError` doesn't need to close the request if the server will do so anyway.

File: http_util_test.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		if readErr != nil {
+			t.Fatalf(readErr.Error())
+		}
+		assert.Equal(t, respErr.Code, CodeUnknown)
+		assert.Equal(t, respErr.Message, err.Error())
+	})
+}
+
+func readConnectErr(resp *http.Response) (*connectWireError, error) {
+	data, err := io.ReadAll(resp.Body)
```
Comment (pkwarren) [inline http_util_test.go:65]: It would be good to add `t.Helper()` here.

File: http_util.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+package connect
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+// WriteError should be called from HTTP middlewares to handle connect errors. It
+// writes a status code encodes a connect error as JSON,  and writes to the response.
```
Comment (pkwarren) [inline http_util.go:23]:
```suggestion
// writes a status code, encodes a connect error as JSON, and writes to the response.
```

File: http_util.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+func WriteError(responseWriter http.ResponseWriter, err error) error {
+	responseWriter.Header().Set(headerContentType, connectUnaryContentTypeJSON)
+	responseWriter.WriteHeader(connectCodeToHTTP(CodeOf(err)))
+	data, marshalErr := json.Marshal(newConnectWireError(err))
```
Comment (pkwarren) [inline http_util.go:27]: nit: Instead of writing to an intermediate []byte, can you use `json.NewEncoder(w).Encode(...)`?

**Review by pkwarren — COMMENTED** [review]
Looks good - just a couple minor things.

**Comment by akshayjshah** [comment]
Awesome, I'm glad you've got some bandwidth to take this on @rhbuf! This is a great start.

Unfortunately, this function only returns valid responses for unary requests using the Connect protocol. If we're going to ship this, we shouldn't make users figure out the protocol in use, the compression scheme, etc - I think we should respond appropriately to unary Connect, streaming Connect, gRPC, and gRPC-Web requests, so that RPC clients get full-fidelity error codes, messages, and details.

**Comment by rhbuf** [comment]
> Awesome, I'm glad you've got some bandwidth to take this on @rhbuf! This is a great start.
> 
> Unfortunately, this function only returns valid responses for unary requests using the Connect protocol. If we're going to ship this, we shouldn't make users figure out the protocol in use, the compression scheme, etc - I think we should respond appropriately to unary Connect, streaming Connect, gRPC, and gRPC-Web requests, so that RPC clients get full-fidelity error codes, messages, and details. Doing that requires access to the request headers.

Roger that, I will come back in a week or so and flesh this out. I've switched directions in my current work and no longer needs this HTTP helper function, but I want to finish what I've started when I have some time.

**Comment by akshayjshah** [comment]
> I've switched directions in my current work and no longer needs this HTTP helper function, but I want to finish what I've started when I have some time.

Don't sweat it! This isn't a "you touch it, you buy it" system :) I'll try to take over and finish up; if I don't get to it and you're free, feel free to pick it up again.

**Comment by akshayjshah** [comment]
@rhbuf and @pkwarren, I took a shot at finishing this up - let me know what you think!

**Review by pkwarren — APPROVED** [review]

**Comment by mattrobenolt** [comment]
🙏 thanks all, this is what I've been waiting for as well.

**Pull request merged**